### PR TITLE
Google Analytics UTM tags for emails and notifications

### PR DESCRIPTION
**Pull request by kuzmany**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Related user documentation PR URL | n
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/3516 https://github.com/mautic/mautic/issues/3106 https://github.com/mautic/mautic/issues/449
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
GA UTM tags support for emails and notifications (web & mobile).
Based on @alanhartless  recommendations from this thread https://github.com/mautic/mautic/issues/3106#issuecomment-267347852

Fund this feature https://mautic-fund.kuzmany.biz/campaign/1/google-analytics-utm-tags-for-channels

#### Steps to test this PR:
1. Apply PR
2. Call php app/console doctrine:schema:update -f
3. Edit email and add utm params in Advanced tab
4. Send email and see update add utm params to links
5. Same steps for notifications

Link should by for example: http://mautic.test/mautic/r/aa84839b312f4ca61e81fd524?ct=YTo1OntzOjY6InNvdXJjZSI7TjtzOjU6ImVtYWlsIjtOO3M6NDoic3RhdCI7czoxNDoieHh4eHh4eHh4eHh4eHgiO3M6NDoibGVhZCI7aTowO3M6NzoiY2hhbm5lbCI7YToxOntzOjU6ImVtYWlsIjtpOjM1O319&utm_medium=mediumtest&utm_name=nametest&utm_content=content+test&utm_source=sourcetest

image

### CODE REVIEW

**Comment by kuzmany** [comment]
@escopecz 
Ready to test.  
I decide create most requested support UTM tags for emails. 
Also added migrations. 
My tests works 👍

File: app/bundles/EmailBundle/Form/Type/EmailUtmTagsType.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        $builder->add(
+            'utmContent',
+            'text',
+            [
+                'label'      => 'mautic.email.campaign_content',
+                'label_attr' => ['class' => 'control-label'],
+                'attr'       => [
+                    'class'    => 'form-control',
+                    //  'preaddon' => 'fa fa-user',
+                    //      'tooltip'  => 'mautic.email.from_name.tooltip',
```
Comment (escopecz) [inline app/bundles/EmailBundle/Form/Type/EmailUtmTagsType.php]: Can we remove the commented lines in this file?

File: app/bundles/PageBundle/Model/TrackableModel.php [hunk 2]
```diff
@@ -87,18 +87,19 @@ protected function getRedirectModel()
      * @param Trackable  $trackable
      * @param array      $clickthrough
      * @param bool|false $shortenUrl   If true, use the configured shortener service to shorten the URLs
+      @param array      $utmTags
      *
```
Comment (escopecz) [inline app/bundles/PageBundle/Model/TrackableModel.php]: `*` is missing here

File: app/migrations/Version20170429092049.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+class Version20170429092049 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        $columns = $schema->getTable($this->prefix.'emails')->getColumns();
+
+        if (array_key_exists('utm_tags', $columns)) {
```
Comment (escopecz) [inline app/migrations/Version20170429092049.php]: The two lines above could be written more easily with
`if ($schema->getTable($this->prefix.'emails')->hasColumn('utm_tags')) {`
It does the same thing, but the code is more easily readable.

File: app/migrations/Version20170429092049.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+class Version20170429092049 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
```
Comment (escopecz) [inline app/migrations/Version20170429092049.php]: This method should be called `preUp`

File: app/migrations/Version20170429092049.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        if (array_key_exists('utm_tags', $columns)) {
+            throw new SkipMigrationException('Schema includes this migration');
+        }
+
+    }
+
+    /**
+     * @param Schema $schema
+     */
+    public function down(Schema $schema)
```
Comment (escopecz) [inline app/migrations/Version20170429092049.php]: And this method should be called `up`

File: app/bundles/PageBundle/Model/RedirectModel.php [hunk 6]
```diff
@@ -74,6 +75,16 @@ public function generateRedirectUrl(Redirect $redirect, $clickthrough = [], $sho
             $shortenUrl
         );
 
+        if (!empty($utmTags)) {
+            $query = parse_url($url, PHP_URL_QUERY);
+            $urlString = http_build_query($utmTags, '', '&');;
```
Comment (escopecz) [inline app/bundles/PageBundle/Model/RedirectModel.php]: one semi-colon is enough ;)

**Comment by kuzmany** [comment]
Fixed

**Comment by escopecz** [comment]
Please look at the conflicts and Travis tests here too.

**Comment by alanhartless** [comment]
@escopecz can you take another look at this and adjust your review?

**Comment by matishaladiwala** [comment]
approved

**Comment by pmgl24** [comment]
Hi Kuzmany, nice work ! Is it possible to remplace utm_name by utm_campaign ? 
utm_name no longer exists since 2012 and Analytics does not understand this variable.

**Comment by kuzmany** [comment]
@pmgl24 this should fix it https://github.com/mautic/mautic/pull/4605. 
Can you test it and  send comment to that PR? 
Thank you for report

**Pull request merged**